### PR TITLE
Completed DatasetFactory and  ImbalancedDatasetSampler

### LINKED ISSUE #2: Imbalanced sampler
- We need a sampler to produce imbalance data from a pytorch dataset. Possibly following this implementation: https://github.com/ufoym/imbalanced-dataset-sampler/blob/master/torchsampler/imbalanced.py
- The sampler should also take as an argument the imbalance ratio (IR). 
- It should also be able to produce a binary dataset (two class only). It could take a tuple of class index for that.

### PR DESCRIPTION
**Pull request by Olivier-tl**
closes #2

### CODE REVIEW

File: datasets/imbalanced_dataset_sampler.py [hunk 1]
```diff
@@ -1,4 +1,41 @@
...
+
+        Args:
+            dataset (torch.utils.data.dataset): Original balanced dataset.
+            imbalance_ratio (int): Imbalance ratio (n_first_class/n_last_class).
+            classes (typing.List[int]): List of classes to sample from.
+        """
+        if imbalance_ratio != 1 and len(classes) == 1:
+            raise ValueError("Can't create an imbalance if only one classe is selected.")
+
+        class_idx = {}
```
Comment (Akssi) [inline datasets/imbalanced_dataset_sampler.py:24]:
```suggestion
        class_idx = defaultdict(list)
```

Comment (Akssi) [reply datasets/imbalanced_dataset_sampler.py:24]: Slightly faster than doing the setdefault in loop
need to import `from collections import default dict`

**Review by Akssi — APPROVED** [review]

File: datasets/imbalanced_dataset_sampler.py (hunk 1 above)
Comment (Olivier-tl) [reply datasets/imbalanced_dataset_sampler.py:24]: Whoa c’est donc ben nice les “suggested change” qu’on peut commit direct 🤯
Et oui c’est plus clean defaultdict, j’voulais faire qqch comme ça mais je savais pas comment 👌🏻

File: datasets/dataset_factory.py [hunk 2]
```diff
@@ -1,17 +1,48 @@
...
-    def create(dataset_name: str, imbalance_ratio: int) -> torch.utils.data.DataLoader:
 
-        # TODO: Create proper dataloader with proper sampler 
+class DatasetFactory:
+    """Constructs the appropriate dataset. Returns train, valid & test splits.
+    """
+    def create(dataset_name: str, imbalance_ratio: int, cache_path: str, validation_split: float,
+               classes: typing.List[int], batch_size: int) -> typing.Tuple[DataLoader, DataLoader, DataLoader]:
+        save_path = os.path.join(cache_path, dataset_name)
+        transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5, ), (0.5, ))])
```
[the remark this answers is not in the document]
Comment (Olivier-tl) [reply datasets/dataset_factory.py:18]: Hmmm good catch you’re right j’vais checker ça, somehow c’est ça pour mnist mais surement pas pour les autres

[the remark this answers is not in the document]
Comment (Olivier-tl) [reply datasets/dataset_factory.py:18]: Actually c'est commun de simplement utiliser mean=0.5 & std=0.5 pour passer de [0, 1] à [-1, 1] sans utiliser le vrai mean/std du training data. https://discuss.pytorch.org/t/normalization-in-the-mnist-example/457/28 

Mais j'pourrais faire un petit script pour le compute

**Review by wwwnev — APPROVED** [review]

File: datasets/dataset_factory.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Olivier-tl) [reply datasets/dataset_factory.py:18]: J'ai ajouté #15 pour ça.

**Comment by Olivier-tl** [comment]
I'm having trouble with the python dependencies on Beluga. I can't install the latest version of torch & torchvision. Sorry if you have conflicts, I pushed my requirements.txt because otherwise github was not able to run the tests.
